diff --git a/common/local/util/sram.sv b/common/local/util/sram.sv
index 8c6c0d3..62f7153 100644
--- a/common/local/util/sram.sv
+++ b/common/local/util/sram.sv
@@ -64,42 +64,40 @@ end
 
   for (genvar k = 0; k<(DATA_WIDTH+63)/64; k++) begin : gen_cut
       // unused byte-enable segments (8bits) are culled by the tool
-      tc_sram_wrapper #(
-        .NumWords(NUM_WORDS),           // Number of Words in data array
-        .DataWidth(64),                 // Data signal width
-        .ByteWidth(32'd8),              // Width of a data byte
-        .NumPorts(32'd1),               // Number of read and write ports
-        .Latency(32'd1),                // Latency when the read data is available
-        .SimInit(SIM_INIT),             // Simulation initialization
-        .PrintSimCfg(1'b0)              // Print configuration
+      prim_ram_1p #(
+        .Depth(NUM_WORDS),       // Number of Words in data array
+        .Width(64),              // Data signal width
+        .DataBitsPerMask(32'd8), // Width of a data byte
+        .MemInitFile(SIM_INIT)   // Simulation initialization
       ) i_tc_sram_wrapper (
           .clk_i    ( clk_i                     ),
           .rst_ni   ( rst_ni                    ),
           .req_i    ( req_i                     ),
-          .we_i     ( we_i                      ),
-          .be_i     ( be_aligned[k*8 +: 8]      ),
+          .write_i  ( we_i                      ),
+          .wmask_i  ( be_aligned[k*8 +: 8]      ),
           .wdata_i  ( wdata_aligned[k*64 +: 64] ),
           .addr_i   ( addr_i                    ),
-          .rdata_o  ( rdata_aligned[k*64 +: 64] )
+          .rdata_o  ( rdata_aligned[k*64 +: 64] ),
+          .cfg_i    ( '0                        ),
+          .cfg_rsp_o(                           )
       );
       if (USER_EN > 0) begin : gen_mem_user
-        tc_sram_wrapper #(
-          .NumWords(NUM_WORDS),           // Number of Words in data array
-          .DataWidth(64),                 // Data signal width
-          .ByteWidth(32'd8),              // Width of a data byte
-          .NumPorts(32'd1),               // Number of read and write ports
-          .Latency(32'd1),                // Latency when the read data is available
-          .SimInit(SIM_INIT),             // Simulation initialization
-          .PrintSimCfg(1'b0)              // Print configuration
+        prim_ram_1p #(
+          .Depth(NUM_WORDS),       // Number of Words in data array
+          .Width(64),              // Data signal width
+          .DataBitsPerMask(32'd8), // Width of a data byte
+          .MemInitFile(SIM_INIT)   // Simulation initialization
         ) i_tc_sram_wrapper_user (
             .clk_i    ( clk_i                     ),
             .rst_ni   ( rst_ni                    ),
             .req_i    ( req_i                     ),
-            .we_i     ( we_i                      ),
-            .be_i     ( be_aligned[k*8 +: 8]      ),
+            .write_i  ( we_i                      ),
+            .wmask_i  ( be_aligned[k*8 +: 8]      ),
             .wdata_i  ( wuser_aligned[k*64 +: 64] ),
             .addr_i   ( addr_i                    ),
-            .rdata_o  ( ruser_aligned[k*64 +: 64] )
+            .rdata_o  ( ruser_aligned[k*64 +: 64] ),
+            .cfg_i    ( '0                        ),
+            .cfg_rsp_o(                           )
         );
       end else begin : gen_mem_user
           assign ruser_aligned[k*64 +: 64] = '0;
diff --git a/core/acc_dispatcher.sv b/core/acc_dispatcher.sv
index c0a707c..ccdd0c0 100644
--- a/core/acc_dispatcher.sv
+++ b/core/acc_dispatcher.sv
@@ -88,7 +88,7 @@ module acc_dispatcher
 
   `include "common_cells/registers.svh"
 
-  import cf_math_pkg::idx_width;
+  import prim_util_pkg::vbits;
 
   /***********************
    *  Common signals     *
@@ -102,7 +102,7 @@ module acc_dispatcher
    **************************/
 
   // Issue accelerator instructions
-  `FF(acc_valid_q, acc_valid_d, '0)
+  `FF(acc_valid_q, acc_valid_d, '0, clk_i, rst_ni)
 
   assign acc_valid_ex_o = acc_valid_q;
   assign acc_valid_d    = ~issue_instr_i.ex.valid &
@@ -140,7 +140,7 @@ module acc_dispatcher
   fu_data_t                                        acc_insn_queue_o;
   logic                                            acc_insn_queue_pop;
   logic                                            acc_insn_queue_empty;
-  logic     [idx_width(InstructionQueueDepth)-1:0] acc_insn_queue_usage;
+  logic     [vbits(InstructionQueueDepth)-1:0] acc_insn_queue_usage;
   logic                                            acc_commit;
   logic     [           CVA6Cfg.TRANS_ID_BITS-1:0] acc_commit_trans_id;
 
@@ -174,13 +174,13 @@ module acc_dispatcher
 
   // Keep track of the instructions that were received by the dispatcher.
   logic [CVA6Cfg.NR_SB_ENTRIES-1:0] insn_pending_d, insn_pending_q;
-  `FF(insn_pending_q, insn_pending_d, '0)
+  `FF(insn_pending_q, insn_pending_d, '0, clk_i, rst_ni)
 
   // Only non-speculative instructions can be issued to the accelerators.
   // The following block keeps track of which transaction IDs reached the
   // top of the scoreboard, and are therefore no longer speculative.
   logic [CVA6Cfg.NR_SB_ENTRIES-1:0] insn_ready_d, insn_ready_q;
-  `FF(insn_ready_q, insn_ready_d, '0)
+  `FF(insn_ready_q, insn_ready_d, '0, clk_i, rst_ni)
 
   always_comb begin : p_non_speculative_ff
     // Maintain state
@@ -261,7 +261,7 @@ module acc_dispatcher
           ],
           rs1     : acc_insn_queue_o.operand_a,
           rs2     : acc_insn_queue_o.operand_b,
-          frm     : fpnew_pkg::roundmode_e'(fcsr_frm_i),
+          frm     : 3'(fcsr_frm_i),
           trans_id: acc_insn_queue_o.trans_id,
           default: '0
       };
@@ -334,7 +334,7 @@ module acc_dispatcher
   // before continuing execution), halt execution while there are pending stores in
   // the accelerator pipeline.
   logic wait_acc_store_d, wait_acc_store_q;
-  `FF(wait_acc_store_q, wait_acc_store_d, '0)
+  `FF(wait_acc_store_q, wait_acc_store_d, '0, clk_i, rst_ni)
 
   // Set on store barrier. Clear when no store is pending.
   assign wait_acc_store_d = (wait_acc_store_q | commit_st_barrier_i) & acc_resp_i.acc_resp.store_pending;
diff --git a/core/cache_subsystem/cva6_icache.sv b/core/cache_subsystem/cva6_icache.sv
index f2ce1e4..c288275 100644
--- a/core/cache_subsystem/cva6_icache.sv
+++ b/core/cache_subsystem/cva6_icache.sv
@@ -408,14 +408,19 @@ module cva6_icache
   );
 
   // generate random cacheline index
-  lfsr #(
-      .LfsrWidth(8),
-      .OutWidth (CVA6Cfg.ICACHE_SET_ASSOC_WIDTH)
+  prim_lfsr #(
+      .LfsrDw     (8),
+      .StateOutDw (CVA6Cfg.ICACHE_SET_ASSOC_WIDTH)
   ) i_lfsr (
-      .clk_i (clk_i),
-      .rst_ni(rst_ni),
-      .en_i  (update_lfsr),
-      .out_o (rnd_way)
+      .clk_i  (clk_i),
+      .rst_ni (rst_ni),
+
+      .seed_en_i (1'b0),
+      .seed_i    ('0),
+      .entropy_i ('0),
+
+      .lfsr_en_i (update_lfsr),
+      .state_o   (rnd_way)
   );
 
 
diff --git a/core/cache_subsystem/wt_dcache_mem.sv b/core/cache_subsystem/wt_dcache_mem.sv
index ad93dd2..af18ed6 100644
--- a/core/cache_subsystem/wt_dcache_mem.sv
+++ b/core/cache_subsystem/wt_dcache_mem.sv
@@ -283,7 +283,7 @@ module wt_dcache_mem
     // In case of an uncached read, return the desired CVA6Cfg.XLEN-bit segment of the most recent AXI read
     assign wr_cl_off     = (wr_cl_nc_i) ? (CVA6Cfg.AxiDataWidth == CVA6Cfg.CLEN) ? '0 :
                            (CVA6Cfg.CheriPresent) ? {{CVA6Cfg.DCACHE_OFFSET_WIDTH-CVA6Cfg.CLEN_ALIGN_BYTES{1'b0}}, wr_cl_off_i[CVA6Cfg.CLEN_ALIGN_BYTES]} :
-                          {{CVA6Cfg.DCACHE_OFFSET_WIDTH-AXI_OFFSET_WIDTH{1'b0}}, wr_cl_off_i[AXI_OFFSET_WIDTH-1:CVA6Cfg.CLEN_ALIGN_BYTES]} :
+                          {{CVA6Cfg.DCACHE_OFFSET_WIDTH-AXI_OFFSET_WIDTH{1'b0}}, wr_cl_off_i[AXI_OFFSET_WIDTH-1-:CVA6Cfg.CLEN_ALIGN_BYTES]} :
                               wr_cl_off_i[CVA6Cfg.DCACHE_OFFSET_WIDTH-1:CVA6Cfg.CLEN_ALIGN_BYTES];
   end else begin : gen_piton_offset
     assign wr_cl_off = wr_cl_off_i[CVA6Cfg.DCACHE_OFFSET_WIDTH-1:3];
diff --git a/core/cache_subsystem/wt_dcache_missunit.sv b/core/cache_subsystem/wt_dcache_missunit.sv
index 76e8db6..cdc7e9c 100644
--- a/core/cache_subsystem/wt_dcache_missunit.sv
+++ b/core/cache_subsystem/wt_dcache_missunit.sv
@@ -198,14 +198,19 @@ module wt_dcache_missunit
   );
 
   // generate random cacheline index
-  lfsr #(
-      .LfsrWidth(8),
-      .OutWidth (CVA6Cfg.DCACHE_SET_ASSOC_WIDTH)
+  prim_lfsr #(
+      .LfsrDw     (8),
+      .StateOutDw (CVA6Cfg.DCACHE_SET_ASSOC_WIDTH)
   ) i_lfsr_inv (
-      .clk_i (clk_i),
-      .rst_ni(rst_ni),
-      .en_i  (update_lfsr),
-      .out_o (rnd_way)
+      .clk_i  (clk_i),
+      .rst_ni (rst_ni),
+
+      .seed_en_i (1'b0),
+      .seed_i    ('0),
+      .entropy_i ('0),
+
+      .lfsr_en_i (update_lfsr),
+      .state_o   (rnd_way)
   );
 
   assign repl_way             = (all_ways_valid) ? rnd_way : inv_way;
@@ -329,7 +334,7 @@ module wt_dcache_missunit
   logic [CVA6Cfg.CLEN_ALIGN_BYTES-1:0] amo_offset;
 
   if (CVA6Cfg.CheriPresent) begin
-    assign amo_offset = ((amo_req_i.size==3'b011) & CVA6Cfg.IS_XLEN64 & CVA6Cfg.CheriPresent) ? amo_req_i.operand_a[CVA6Cfg.CLEN_ALIGN_BYTES-1:0] + 7 : 
+    assign amo_offset = ((amo_req_i.size==3'b011) & CVA6Cfg.IS_XLEN64 & CVA6Cfg.CheriPresent) ? amo_req_i.operand_a[CVA6Cfg.CLEN_ALIGN_BYTES-1:0] + 7 :
                         ((amo_req_i.size==3'b010) & CVA6Cfg.IS_XLEN64) ? amo_req_i.operand_a[CVA6Cfg.CLEN_ALIGN_BYTES-1:0] + 3 :
                         ((amo_req_i.size==3'b001))  ? amo_req_i.operand_a[CVA6Cfg.CLEN_ALIGN_BYTES-1:0] + 1 :
                                                       amo_req_i.operand_a[CVA6Cfg.CLEN_ALIGN_BYTES-1:0];
diff --git a/core/cva6_mmu/cva6_shared_tlb.sv b/core/cva6_mmu/cva6_shared_tlb.sv
index d7cfbfe..ca0f02f 100644
--- a/core/cva6_mmu/cva6_shared_tlb.sv
+++ b/core/cva6_mmu/cva6_shared_tlb.sv
@@ -433,14 +433,19 @@ module cva6_shared_tlb #(
       .empty_o(all_ways_valid)
   );
 
-  lfsr #(
-      .LfsrWidth(8),
-      .OutWidth ($clog2(SHARED_TLB_WAYS))
+  prim_lfsr #(
+      .LfsrDw     ( 8                       ),
+      .StateOutDw ( $clog2(SHARED_TLB_WAYS) )
   ) i_lfsr (
-      .clk_i (clk_i),
-      .rst_ni(rst_ni),
-      .en_i  (update_lfsr),
-      .out_o (rnd_way)
+      .clk_i  (clk_i),
+      .rst_ni (rst_ni),
+
+      .seed_en_i (1'b0),
+      .seed_i    ('0),
+      .entropy_i ('0),
+
+      .lfsr_en_i (update_lfsr),
+      .state_o   (rnd_way)
   );
 
   ///////////////////////////////////////////////////////
diff --git a/core/ex_stage.sv b/core/ex_stage.sv
index 61550f5..10f96ad 100644
--- a/core/ex_stage.sv
+++ b/core/ex_stage.sv
@@ -308,10 +308,10 @@ module ex_stage
     rs2_forwarding = rs2_forwarding_i[0];
 
     if (CVA6Cfg.SuperscalarEn) begin
-      if (one_cycle_select[1]) begin
-        one_cycle_data = fu_data_i[1];
-        rs1_forwarding = rs1_forwarding_i[1];
-        rs2_forwarding = rs2_forwarding_i[1];
+      if (one_cycle_select[CVA6Cfg.SuperscalarEn]) begin
+        one_cycle_data = fu_data_i[CVA6Cfg.SuperscalarEn];
+        rs1_forwarding = rs1_forwarding_i[CVA6Cfg.SuperscalarEn];
+        rs2_forwarding = rs2_forwarding_i[CVA6Cfg.SuperscalarEn];
       end
     end
   end
@@ -419,8 +419,8 @@ module ex_stage
   always_comb begin
     mult_data = mult_valid_i[0] ? fu_data_i[0] : '0;
     if (CVA6Cfg.SuperscalarEn) begin
-      if (mult_valid_i[1]) begin
-        mult_data = fu_data_i[1];
+      if (mult_valid_i[CVA6Cfg.SuperscalarEn]) begin
+        mult_data = fu_data_i[CVA6Cfg.SuperscalarEn];
       end
     end
   end
@@ -456,8 +456,8 @@ module ex_stage
       always_comb begin
         fpu_data = fpu_valid_i[0] ? fu_data_i[0] : '0;
         if (CVA6Cfg.SuperscalarEn) begin
-          if (fpu_valid_i[1]) begin
-            fpu_data = fu_data_i[1];
+          if (fpu_valid_i[CVA6Cfg.SuperscalarEn]) begin
+            fpu_data = fu_data_i[CVA6Cfg.SuperscalarEn];
           end
         end
       end
@@ -560,9 +560,9 @@ module ex_stage
     lsu_tinst = tinst_i[0];
 
     if (CVA6Cfg.SuperscalarEn) begin
-      if (lsu_valid_i[1]) begin
-        lsu_data  = fu_data_i[1];
-        lsu_tinst = tinst_i[1];
+      if (lsu_valid_i[CVA6Cfg.SuperscalarEn]) begin
+        lsu_data  = fu_data_i[CVA6Cfg.SuperscalarEn];
+        lsu_tinst = tinst_i[CVA6Cfg.SuperscalarEn];
       end
     end
   end
@@ -654,8 +654,8 @@ module ex_stage
     always_comb begin
       cvxif_data = x_valid_i[0] ? fu_data_i[0] : '0;
       if (CVA6Cfg.SuperscalarEn) begin
-        if (x_valid_i[1]) begin
-          cvxif_data = fu_data_i[1];
+        if (x_valid_i[CVA6Cfg.SuperscalarEn]) begin
+          cvxif_data = fu_data_i[CVA6Cfg.SuperscalarEn];
         end
       end
     end
diff --git a/core/frontend/instr_queue.sv b/core/frontend/instr_queue.sv
index 98b1481..7e9648b 100644
--- a/core/frontend/instr_queue.sv
+++ b/core/frontend/instr_queue.sv
@@ -530,8 +530,8 @@ module instr_queue
       .pop_i     (pop_address)
   );
 
-  unread i_unread_branch_mask (.d_i(|branch_mask_extended));
-  unread i_unread_fifo_pos (.d_i(|fifo_pos_extended));  // we don't care about the lower signals
+  logic unused_branch_and_fifo;
+  assign unused_branch_and_fifo = |branch_mask_extended | |fifo_pos_extended;
 
   if (CVA6Cfg.RVC) begin : gen_pc_q_with_c
     always_ff @(posedge clk_i or negedge rst_ni) begin
diff --git a/core/id_stage.sv b/core/id_stage.sv
index e819ffa..05f9b77 100644
--- a/core/id_stage.sv
+++ b/core/id_stage.sv
@@ -166,7 +166,7 @@ module id_stage #(
   logic                                               commit_redirect_q;
 
   assign int_mode_decode[0] = int_mode_q;
-  for (genvar i = 1; i <= CVA6Cfg.NrIssuePorts; i++) begin
+  for (genvar i = 1; i < CVA6Cfg.NrIssuePorts; i++) begin
     assign int_mode_decode[i] = int_mode_decode_o[i-1];
   end
 
diff --git a/core/include/ariane_pkg.sv b/core/include/ariane_pkg.sv
index 1dd5b39..a2ddc6e 100644
--- a/core/include/ariane_pkg.sv
+++ b/core/include/ariane_pkg.sv
@@ -864,6 +864,7 @@ package ariane_pkg;
           default: ;  // Do nothing
         endcase
       end
+      default: return 16'b0;
     endcase
     return 16'b0;
   endfunction
diff --git a/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv b/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv
index f0a32ae..3ca9969 100644
--- a/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv
+++ b/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv
@@ -12,8 +12,8 @@ package cva6_config_pkg;
 
   localparam CVA6ConfigXlen = 64;
 
-  localparam CVA6ConfigRVF = 1;
-  localparam CVA6ConfigRVD = 1;
+  localparam CVA6ConfigRVF = 0;
+  localparam CVA6ConfigRVD = 0;
   localparam CVA6ConfigF16En = 0;
   localparam CVA6ConfigF16AltEn = 0;
   localparam CVA6ConfigF8En = 0;
@@ -32,7 +32,7 @@ package cva6_config_pkg;
   localparam CVA6ConfigRVZcherihybrid = 1;
   localparam CVA6ConfigCheriCapTagWidth = 1;
   localparam CVA6ConfigRVFI_DII = 0;
-  localparam CVA6ConfigDIIIDLEN = 0;
+  localparam CVA6ConfigDIIIDLEN = 1;
 
   localparam CVA6ConfigAxiIdWidth = 4;
   localparam CVA6ConfigAxiAddrWidth = 64;
diff --git a/core/issue_read_operands.sv b/core/issue_read_operands.sv
index 914272f..670d2c6 100644
--- a/core/issue_read_operands.sv
+++ b/core/issue_read_operands.sv
@@ -693,10 +693,12 @@ module issue_read_operands
           forward_rs2[0] = 1'b0;
           stall_rs2[0]   = 1'b0;
         end
-        if (OPERANDS_PER_INSTR == 3 && ~x_issue_resp_i.register_read[2]) begin
-          forward_rs3[0] = 1'b0;
-          stall_rs3[0]   = 1'b0;
-        end
+        //TODO if we ever need CvxifEn
+        //if (OPERANDS_PER_INSTR == 3 && ~x_issue_resp_i.register_read[2]) begin
+        //  forward_rs3[0] = 1'b0;
+        //  stall_rs3[0]   = 1'b0;
+        //end
+        $error("CvxifEn not supported.");
       end
       stall_raw[0] = x_transaction_rejected ? 1'b0 : stall_rs1[0] || stall_rs2[0] || stall_rs3[0];
     end
diff --git a/core/store_unit.sv b/core/store_unit.sv
index 323fdf5..e55b925 100644
--- a/core/store_unit.sv
+++ b/core/store_unit.sv
@@ -126,6 +126,7 @@ module store_unit
             4'b1101: data_tmp[CVA6Cfg.CLEN-1:CVA6Cfg.XLEN] = {data[23:0], data[63:24]};
             4'b1110: data_tmp[CVA6Cfg.CLEN-1:CVA6Cfg.XLEN] = {data[15:0], data[63:16]};
             4'b1111: data_tmp[CVA6Cfg.CLEN-1:CVA6Cfg.XLEN] = {data[7:0], data[63:8]};
+            default: data_tmp[CVA6Cfg.CLEN-1:CVA6Cfg.XLEN] = '0;
           endcase
         end
         if (CVA6Cfg.IS_XLEN64) begin
