diff --git a/common/local/util/sram.sv b/common/local/util/sram.sv
index 62f7153..2a02373 100644
--- a/common/local/util/sram.sv
+++ b/common/local/util/sram.sv
@@ -23,7 +23,7 @@ module sram #(
     parameter USER_WIDTH = 1,
     parameter USER_EN    = 0,
     parameter NUM_WORDS  = 1024,
-    parameter SIM_INIT   = "none",
+    parameter SIM_INIT   = "",
     parameter OUT_REGS   = 0     // enables output registers in FPGA macro (read lat = 2)
 )(
    input  logic                          clk_i,
diff --git a/common/local/util/sram_cache.sv b/common/local/util/sram_cache.sv
index 7ca7559..01f90e6 100644
--- a/common/local/util/sram_cache.sv
+++ b/common/local/util/sram_cache.sv
@@ -23,7 +23,7 @@ module sram_cache #(
     parameter USER_WIDTH = 1,
     parameter USER_EN    = 0,
     parameter NUM_WORDS  = 1024,
-    parameter SIM_INIT   = "none",
+    parameter SIM_INIT   = "",
     parameter BYTE_ACCESS = 1,
     parameter TECHNO_CUT = 0,
     parameter OUT_REGS   = 0     // enables output registers in FPGA macro (read lat = 2)
diff --git a/core/cva6.sv b/core/cva6.sv
index 4522765..19dc069 100644
--- a/core/cva6.sv
+++ b/core/cva6.sv
@@ -1438,7 +1438,7 @@ module cva6
         .clk_i             (clk_i),
         .rst_ni            (rst_ni),
         // I$
-        .icache_en_i       (icache_en_csr),
+        .icache_en_i       (1'b0),
         .icache_flush_i    (icache_flush_ctrl_cache),
         .icache_miss_o     (icache_miss_cache_perf),
         .icache_areq_i     (icache_areq_ex_cache),
@@ -1446,7 +1446,7 @@ module cva6
         .icache_dreq_i     (icache_dreq_if_cache),
         .icache_dreq_o     (icache_dreq_cache_if),
         // D$
-        .dcache_enable_i   (dcache_en_csr_nbdcache),
+        .dcache_enable_i   (1'b0),
         .dcache_flush_i    (dcache_flush_ctrl_cache),
         .dcache_flush_ack_o(dcache_flush_ack_cache_ctrl),
         // to commit stage
@@ -1810,7 +1810,8 @@ module cva6
     if (~rst_ni) begin
       cycles <= 0;
     end else begin
-      byte mode = "";
+      byte mode;
+      mode = "";
       if (CVA6Cfg.DebugEn && debug_mode) mode = "D";
       else begin
         case (priv_lvl)
diff --git a/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv b/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv
index 3ca9969..683329b 100644
--- a/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv
+++ b/core/include/cv64a6_imafdchzcheri_sv39_config_pkg.sv
@@ -141,11 +141,11 @@ package cva6_config_pkg;
       NonIdempotentAddrBase: 1024'({64'b0, 64'b0}),
       NonIdempotentLength: 1024'({64'b0, 64'b0}),
       NrExecuteRegionRules: unsigned'(3),
-      ExecuteRegionAddrBase: 1024'({64'h8000_0000, 64'h1_0000, 64'h0}),
-      ExecuteRegionLength: 1024'({64'h40000000, 64'h10000, 64'h1000}),
+      ExecuteRegionAddrBase: 1024'({64'h8000_0000, 64'h10_0000, 64'h0}),
+      ExecuteRegionLength: 1024'({64'h40000000, 64'h10_0000, 64'h1000}),
       NrCachedRegionRules: unsigned'(1),
-      CachedRegionAddrBase: 1024'({64'h8000_0000}),
-      CachedRegionLength: 1024'({64'h40000000}),
+      CachedRegionAddrBase: 1024'({64'hE000_0000}),
+      CachedRegionLength: 1024'({64'h1000_0000}),
       MaxOutstandingStores: unsigned'(7),
       DebugEn: bit'(1),
       AxiBurstWriteEn: bit'(1),
