diff --git a/ibex_core.core b/ibex_core.core
index 15fb327..4fea69a 100644
--- a/ibex_core.core
+++ b/ibex_core.core
@@ -42,10 +42,6 @@ filesets:
     files:
       - lint/verilator_waiver.vlt: {file_type: vlt}
 
-  files_check_tool_requirements:
-    depend:
-     - lowrisc:tool:check_tool_requirements
-
 parameters:
   RVFI:
     datatype: bool
@@ -167,7 +163,6 @@ targets:
     filesets:
       - tool_verilator ? (files_lint_verilator)
       - files_rtl
-      - files_check_tool_requirements
     toplevel: ibex_core
     parameters:
       - tool_vivado ? (FPGA_XILINX=true)
diff --git a/ibex_top.core b/ibex_top.core
index 38a279d..36a351b 100644
--- a/ibex_top.core
+++ b/ibex_top.core
@@ -30,10 +30,6 @@ filesets:
     files:
       - lint/verilator_waiver.vlt: {file_type: vlt}
 
-  files_check_tool_requirements:
-    depend:
-     - lowrisc:tool:check_tool_requirements
-
 parameters:
   RVFI:
     datatype: bool
@@ -155,7 +151,6 @@ targets:
     filesets:
       - tool_verilator ? (files_lint_verilator)
       - files_rtl
-      - files_check_tool_requirements
     toplevel: ibex_top
     parameters:
       - tool_vivado ? (FPGA_XILINX=true)
