diff --git a/dv/sva/pwrmgr_rstreqs_sva_if.sv.tpl b/dv/sva/pwrmgr_rstreqs_sva_if.sv.tpl
index bf64e38..d88abeb 100644
--- a/dv/sva/pwrmgr_rstreqs_sva_if.sv.tpl
+++ b/dv/sva/pwrmgr_rstreqs_sva_if.sv.tpl
@@ -38,6 +38,9 @@ interface pwrmgr_rstreqs_sva_if
 % elif topname == "darjeeling":
   // Darjeeling has a 16x clock ratio between the slow and fast clocks. (1GHz / 62.5MHz = 16)
   localparam int FastToSlowFreqRatio = 16;
+% elif topname == "mocha":
+  // Mocha has a 20x clock ratio between the slow (io/aon) clocks and the fast system clock. (1000MHz / 50MHz = 20)
+  localparam int FastToSlowFreqRatio = 20;
 % else:
 <% assert False, f"Unsupported top: {topname}" %>\
 % endif
